### PR TITLE
dns: destroy/reinitialize c-ares channel on ARES_ECONNREFUSED

### LINKED ISSUE #4543: dns: DnsResolverImpl keeps using a "broken" c-ares channel
We have a `STRICT_DNS` type of a cluster defined in bootstrap config. In one of our test Pods, the membership count of this cluster became zero. This is understandable because the DNS resolution might have resulted in zero hosts. However this remained like this for quite a long time and after killing the container, Envoy is able to successfully resolve the DNS. 

I have taken debug logs when Envoy is not able to resolve this. I see the following line

"source/common/network/dns_impl.cc:118] `DNS request timed out 4 times`",,

And I see these lines repeatedly
"source/common/network/dns_impl.cc:147] Setting DNS resolution timer for 0 milliseconds",,
"source/common/network/dns_impl.cc:147] Setting DNS resolution timer for 0 milliseconds",,
"source/common/network/dns_impl.cc:147] Setting DNS resolution timer for 0 milliseconds",,
"source/common/network/dns_impl.cc:147] Setting DNS resolution timer for 0 milliseconds",,
"source/common/network/dns_impl.cc:147] Setting DNS resolution timer for 0 milliseconds",,
"source/common/network/dns_impl.cc:147] Setting DNS resolution timer for 0 milliseconds",,
"source/common/network/dns_impl.cc:147] Setting DNS resolution timer for 0 milliseconds",,
"source/common/network/dns_impl.cc:147] Setting DNS resolution timer for 22 milliseconds"

So at this point I am not very clear if it is Envoy issue or container DNS issue - as container restart resolved the issue.
Has any one seen similar issues with DNS? and another question is it the DNS resolution timer behaviour correct in the sense it is trying to resolve 0 milliseconds?

### PR DESCRIPTION
**Pull request by junr03**
Description: this PR adds logic to the DnsResolverImpl to destroy and re-initialize its c-ares channel under certain circumstances. A better option would require work in c-ares https://github.com/c-ares/c-ares/issues/301.
Risk Level: med changes in low-level DNS resolution.
Testing: unit tests

Fixes #4543 

Signed-off-by: Jose Nino <jnino@lyft.com>

### CODE REVIEW

**Comment by junr03** [comment]
opening as draft until I write tests monday.

cc @htuch as the original writer.

File: source/common/network/dns_impl.cc [hunk 1]
```diff
@@ -171,6 +183,7 @@ void DnsResolverImpl::updateAresTimer() {
 }
 
 void DnsResolverImpl::onEventCallback(int fd, uint32_t events) {
+  ENVOY_LOG(error, "onEventCallback");
```
Comment (mattklein123) [inline source/common/network/dns_impl.cc]: del

File: source/common/network/dns_impl.cc [hunk 2]
```diff
@@ -203,8 +216,23 @@ ActiveDnsQuery* DnsResolverImpl::resolve(const std::string& dns_name,
...
+  if (dirty_channel_) {
+    ares_destroy(channel_);
+
+    ares_options options{};
+    int optmask = 0;
+
+    if (use_tcp_for_dns_lookups_) {
+      optmask |= ARES_OPT_FLAGS;
+      options.flags |= ARES_FLAG_USEVC;
+    }
```
Comment (mattklein123) [inline source/common/network/dns_impl.cc]: This should be factored into a helper function or moved into `initializeChannel`

File: source/common/network/dns_impl.cc [hunk 3]
```diff
@@ -203,8 +216,23 @@ ActiveDnsQuery* DnsResolverImpl::resolve(const std::string& dns_name,
   // TODO(hennna): Add DNS caching which will allow testing the edge case of a
   // failed initial call to getHostByName followed by a synchronous IPv4
   // resolution.
+
+  // @see DnsResolverImpl::PendingResolution::onAresGetAddrInfoCallback for why this is done.
+  if (dirty_channel_) {
```
Comment (mattklein123) [inline source/common/network/dns_impl.cc]: You need to set dirty_channel_ = false; in this block.

File: source/common/network/dns_impl.h [hunk 4]
```diff
@@ -93,6 +95,8 @@ class DnsResolverImpl : public DnsResolver, protected Logger::Loggable<Logger::I
   Event::Dispatcher& dispatcher_;
   Event::TimerPtr timer_;
   ares_channel channel_;
+  bool dirty_channel_{};
+  bool use_tcp_for_dns_lookups_;
```
Comment (mattklein123) [inline source/common/network/dns_impl.h]: nit: const

**Comment by junr03** [comment]
Whoops did not mean for you to review my half baked code. Thanks for the comments though, I agree with all of them @mattklein123. Will update and write tests Monday.

File: source/common/network/dns_impl.cc [hunk 5]
```diff
@@ -83,6 +89,19 @@ void DnsResolverImpl::PendingResolution::onAresGetAddrInfoCallback(int status, i
...
+    // broken. Mark the channel dirty so that it is destroyed and reinitialized on a subsequent call
+    // to DnsResolver::resolve(). The optimal solution would be for c-ares to reinitialize the
+    // channel, and not have Envoy track side effects.
+    // context: https://github.com/envoyproxy/envoy/issues/4543 and
+    // https://github.com/c-ares/c-ares/issues/301.
+    //
+    // The channel cannot be destroyed and reinitialized here because that leads to a c-ares
+    // segfault.
+    if (status == ARES_ECONNREFUSED) {
+      parent_.dirty_channel_ = true;
```
Comment (junr03) [inline source/common/network/dns_impl.cc]: I thought about this for a bit and debated between returning here, vs. allowing execution to continue and eventually call the callback_.

I ended up going for allowing the callback for a couple reasons:
1.  Ease of testing, as having the callback allows for nice event loop triggers and a way to test assertions on the address_list.
2. It preserves previous behavior where every time resolution completed, whether successfully or not (with the exception of destruction), the callback_ was called. Obviously after fallbacks were exhausted, and if the query was not cancelled.

wdyt?

Comment (mattklein123) [reply source/common/network/dns_impl.cc]: Yeah I agree we need to call the callback. As you already pointed out this is related to the other issue, so let's just fix the callbacks to expose errors in a future change.

File: test/common/network/dns_impl_test.cc [hunk 6]
```diff
@@ -582,6 +590,64 @@ TEST_P(DnsImplTest, CallbackException) {
...
+
+  dispatcher_->run(Event::Dispatcher::RunType::Block);
+  // The c-ares channel should be dirty because the TestDnsServer replied with return code REFUSED;
+  // This test, and the way the TestDnsServerQuery is setup, relies on the fact that Envoy's
+  // c-ares channel is configured **without** the ARES_FLAG_NOCHECKRESP flag. This causes c-ares to
+  // discard packets with REFUSED, and thus Envoy receives ARES_ECONNREFUSED due to the code here:
+  // https://github.com/c-ares/c-ares/blob/d7e070e7283f822b1d2787903cce3615536c5610/ares_process.c#L654
+  // If that flag needs to be set, or c-ares changes its handling this test will need to be updated
+  // to create another condition where c-ares invokes onAresGetAddrInfoCallback with status ==
+  // ARES_ECONNREFUSED.
```
Comment (mattklein123) [inline test/common/network/dns_impl_test.cc]: Agreed this is a little fragile but I think it's OK for now. Nice work figuring out a way to test this!

**Pull request merged**